### ISSUE TITLE
Rename packages to match JUnit Pioneer

### ISSUE THREAD

**Issue by nicolaiparlog** [junit-pioneer/junit-pioneer#24]
Current package prefix is `org.codefx.junit.io`. The new prefix is not decided upon yet but the domain `junit-pioneer.org` belongs to the project and might be used.

**Comment by nicolaiparlog** [comment]
My last coherent thoughts on the matter were:

> To me, this is the most difficult decision at the moment. What is our general opinion on the matter? Would we use `org.junit-pioneer` if it were legal?
>
> I have no strong opinion on that but I guess that would be my default choice. If so, we obviously need to get rid of the dash and there are seem to be three choices for replacements:
>
> 0. "." -> `org.junit.pioneer`
> 1. "" -> `org.junitpioneer`
> 2. "_" -> `org.junit_pioneer`
> 3. "$" -> `org.junit$pioneer`
> 
> 0. has been ruled out in junit-team/junit5#675. 1. is a solid choice but a little boring. I think we agreed that "_" is ugly. Re 3. the JLS says:
>
>> The $ character should be used only in mechanically generated source code or, rarely, to access pre-existing names on legacy systems.
>
> Other than that I have to admit that I actually like it.
> 
> Other routes would be to start with `junit.pioneer` or `pioneer` - I'm not a fan of either. The former is too close to how JUnit 4 named packages (used to start with `junit`) and the latter loses the connection to JUnit.
>
> So... my current choices are:
> 
> A. `org.junitpioneer` (safe and known)
> B. `org.junit$pioneer` (weird and bold, or just stupid)

**Comment by jbduncan** [comment]
`org.junitpioneer` looks like the most sensible option to me.

Although, if `org.junit-pioneer` were legal (I admit I don't see yet why it's not legal), I'd probably vote for that. :)

**Comment by nicolaiparlog** [comment]
> I admit I don't see yet why it's not legal

Package names can have no dashes. Only special characters `_` and `$` are allowed.

**Comment by nicolaiparlog** [comment]
@smoyer64 What is your opinion on this? A or B?

> A. org.junitpioneer (safe and known)
> B. org.junit$pioneer (weird and bold, or just stupid)

**Comment by smoyer64** [comment]
I think I'm going to have to go with the safe and known option (but could live with either)

**Comment by nicolaiparlog** [comment]
Well if you could live with either than I want to do the crazy thing! (Important lesson for our collaboration: Given the choice, I can not be trusted to avoid the crazy one. 😉)

**Comment by smoyer64** [comment]
Woo hoo - "I am a wild and crazy guy"

**Comment by nicolaiparlog** [comment]
Look at my hair - apparently I need that microrevolution. :wink:

**Comment by smoyer64** [comment]
Sorry ... that's a reference that might be a bit old for this crowd ;) - https://www.youtube.com/watch?v=dpFNC7puU2M

**Comment by nicolaiparlog** [comment]
> that's a reference that might be a bit old for this crowd

Indeed. 😂 I had to watch [the next one](https://www.youtube.com/watch?v=YGv6uo89yMY) for the quote, though (or did I miss it in the one you linked?).

**Comment by smoyer64** [comment]
Hmm ... I guess I couldn't read their lips as well as I thought (in a meeting with the audio muted).

**Comment by marcphilipp** [comment]
While the `$` looks very courageous, it might confuse people and tools.

I agree with what @jbduncan wrote:

> `org.junitpioneer` looks like the most sensible option to me.

I think we should listen to our users (when they share my opinion anyway)... 😉

@nicolaiparlog Would you be offended if we reopened this issue to discuss?

**Comment by mkobit** [comment]
FWIW I was confused by the `$` in the package name in https://github.com/junit-pioneer/junit-pioneer/pull/63#issuecomment-331030489

**Comment by sbrannen** [comment]
As we all know, naming things is one of the hardest tasks in computer science.

But in any case, usability and utility should take precedence over coolness/craziness whenever selecting a name.

In particular, the package name and automatic module name should follow the guidelines of the specifications as well as conventional wisdom (and common expectations).

The inclusion of a `$` in both of those violates those principles.

Although the `$` technically works (in terms of permissible syntax), that doesn't mean we _should_ use it just because we _could_ use it.

In fact, I am strongly against the inclusion of a `$` in either the package name or the automatic module name.

- People familiar with Java will automatically think something is broken when they encounter it.
  - It throws me off mentally every single time I see it and not in a pleasant way.
- The `$` is supposed to imply generated code, which is not the case here.
- The `$` has special semantics in numerous languages (e.g., Groovy, Bash, etc.).
  - This can potentially make it rather annoying in cases where variable expansion might occur, thereby requiring users to escape the `$` in build scripts (Gradle/Groovy), at the command line, in shell scripts, etc.

In summary, although some people may find the `$` _cool_, it may well turn out that the community finds it very _uncool_ over time.

With that in mind, I think the usability and utility factors far outweigh the coolness factor.

I therefore vote for `org.junitpioneer` as the package name (with the automatic module name modified accordingly), and I think obtaining the `org.junitpioneer` domain name would be wise as well.

**Comment by nicolaiparlog** [comment]
# Confusion 

> FWIW I was confused by the `$` in the package name in https://github.com/junit-pioneer/junit-pioneer/pull/63#issuecomment-331030489

> In particular, the package name and automatic module name should follow [...] conventional wisdom (and common expectations).

> It throws me off mentally every single time I see it and not in a pleasant way.

I would not follow conventional wisdom and common expectations if they don't add any value (on their own). And unless confusion makes it more likely that you make a mistake, I don't think it's a good argument. Many things are confusing until you get used to them (`method::references` anyone?).

# Specification

> In particular, the package name and automatic module name should follow the guidelines of the specifications [...].

That counts! I agree that `junit$pioneer`'s deviation from the specification can be seen as a problem. To clarify, here's a quote from [the JLS on identifiers](https://docs.oracle.com/javase/specs/jls/se8/html/jls-3.html#jls-Identifier) (addition and emphasis mine):

> The "Java letters" [which are to be used for package names] include uppercase and lowercase ASCII Latin letters A-Z (\u0041-\u005a), and a-z (\u0061-\u007a), and, for historical reasons, the ASCII underscore (_, or \u005f) and dollar sign ($, or \u0024). **The $ sign should be used only in mechanically generated source code or, rarely, to access pre-existing names on legacy systems.**

That's pretty clear. But I'm willing to ignore it if `$` has no negative practical consequences.

# Just To Be Cool

And yes, just to have a cool name. :sunglasses: Names matter, and having a good/interesting one counts for something. (I'm sure you guys were happy when you discovered that Jupiter aligns so nicely with JUnit 5.)

Also, as someone constantly looking stuff up to make sure I get it right, names like `org.springframework` "and 1000s of similar cases" do actually annoy me. "Class path" vs "classpath", "file system" vs "filesystem", ["run time" vs "runtime" vs "run-time"](https://twitter.com/nipafx/status/919893075509555200), "Oracle JDK" vs "OpenJDK", ... Why add one to the list?

None of the arguments I addressed so far convince me and I get the impression that they are mainly based in an aesthetic aversion to that uncommon name. That's totally fair, but if that's the root of the discussion, let's have one about the aesthetics.

# Negative Practical Consequences

Here's where actual consequences (as opposed to convention and taste) enter the conversation.

> The $ has special semantics in numerous languages (e.g., Groovy, Bash, etc.).
>
> This can potentially make it rather annoying in cases where variable expansion might occur, thereby requiring users to escape the $ in build scripts (Gradle/Groovy), at the command line, in shell scripts, etc.

That's actually a pretty good argument. I'm not sure it will occur all that often, but even if not, the name's not a good reason to make life harder for people who are going down that road.

# Resolution?

The argument that `junit$pioneer` goes against the JLS' recommendation holds, but I'm willing to ignore that and I don't buy into the other "convention-based" lines of thought. The technical problems have me half convinced.

But as I see it, I'm the only one on this side of the argument right? I already assumed that Steve would not have picked that name, Marc and Sam obviously don't like it either, and Jonathan and Mike also  spoke out against it, so... it I'm inclined to give in to peer pressure.

But not yet. Let's see [what the public thinks](https://twitter.com/nipafx/status/1003547161248845824)! (Retweets appreciated.)

**Comment by nicolaiparlog** [comment]
PS: Btw, I'm sure the public will back up the removal of `$`, so I don't have high hopes. :wink:

**Comment by sbrannen** [comment]
I also considered publishing a poll on Twitter.

So.... yep.... let's see how people vote.

**Comment by smoyer64** [comment]
@nicolaiparlog 

>  I already assumed that Steve would not have picked that name,

IIRC (from our original discussion), I've admitted that I don't like names just appended together ... perhaps I'm old and my eyes just don't pick the words out as well?  But I don't like _ either so I specifically stated that  I didn't have a preference between those two separators.

My first choice would still be org.junit-pioneer (but dashes aren't allowed but this should still be the groupId).  My second choice would have been org.junit.pioneer but that's a sub-domain of the official JUnit project.  And so I'm still of the position that I simply don't care.  I think there's one other argument (that I haven't tested - how do IDEs feel about a package name with $ in it?

**Comment by nicolaiparlog** [comment]
> how do IDEs feel about a package name with $ in it?

All the tools I use work fine for my use cases: IntelliJ, Gradle, JUnit, Mockito, javac, jar, java, javadoc...

**Comment by nicolaiparlog** [comment]
The [results](https://twitter.com/nipafx/status/1003547161248845824) are in:

> 4%: Keep $, it looks great!
> 51%: Remove $, it's ugly
> 35%: Remove $, it's trouble
> 10%: Don't care, I auto-import
>
> 171 votes

Interestingly enough, even though I explicitly asked for this, none of the 60 people who thought `$` would cause trouble could add any concrete example of where it might.

But be that as it may, I'm not one to push through something as unimportant as a name. So what's it gonna be, `org.junitpioneer` or `org.junit_pioneer`? Both are ugly but the latter is at least readable, so I prefer that one.

**Comment by sormuras** [comment]
`org.junitpioneer` +1

**Comment by sbrannen** [comment]
`org.junitpioneer` +1

**Comment by nicolaiparlog** [comment]
Now you're just trolling me. :rofl:

@smoyer64: I let you have the last word and then prepare a PR with the new package name.

**Comment by smoyer64** [comment]
Thanks (I guess) ... you're letting me be the bad guy?  After stating that I didn't care?

I hate underscores in package names, so I vote that Java 11 allows dashes in package names.

**Comment by seanf** [comment]
What ever happened about registering the domain **junitpioneer.org** (as suggested by @sbrannen)?

**Comment by nicolaiparlog** [comment]
Domains cost money and I'm not made of it. :wink:

**Comment by smoyer64** [comment]
Our groupId is org.junit-pioneer ... are we proposing that that change as well?

**Comment by sbrannen** [comment]
Yes, we should change the group ID accordingly.
